### PR TITLE
Switch to Proxy-less SyncCommitteeCached

### PR DESCRIPTION
**Pull request by dapplion**
Blocked by https://github.com/ChainSafe/lodestar/pull/3728

**Motivation**

With SSZ refactor we have the option to drop the CachedBeaconState Proxies and added flat vectors.
- Performance benchmarks have shown that going through Proxies is significantly slower that regular objects
- Memory benchmarks have shown that data duplication is significant.

This PR moves SyncCommittee caches from Proxied CachedBeaconState to epochCtx. Overall the logic is very similar, but this re-organization allows to drop the root Proxy eventually

**Description**

- Switch to Proxy-less SyncCommitteeCached

### CODE REVIEW

File: packages/beacon-state-transition/src/util/genesis.ts [hunk 1]
```diff
@@ -89,7 +89,7 @@ export function getGenesisBeaconState(
   state.eth1Data = genesisEth1Data;
   state.randaoMixes = randaoMixes;
 
-  return createCachedBeaconState(config, state);
+  return createCachedBeaconState(config, state, {skipSyncCommitteeCache: true});
```
Comment (wemeetagain) [inline packages/beacon-state-transition/src/util/genesis.ts:100]: Why is this added here? Don't we want the cache populated if the chain is initialized with ALTAIR_FORK_EPOCH=0 ?

File: packages/lodestar/src/chain/lightClient/proofs.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
-import {allForks} from "@chainsafe/lodestar-types";
+import {altair} from "@chainsafe/lodestar-types";
 import {TreeBacked} from "@chainsafe/ssz";
 import {FINALIZED_ROOT_GINDEX} from "@chainsafe/lodestar-params";
 import {SyncCommitteeWitness} from "./types";
 
-export function getSyncCommitteesWitness(state: TreeBacked<allForks.BeaconState>): SyncCommitteeWitness {
+export function getSyncCommitteesWitness(state: TreeBacked<altair.BeaconState>): SyncCommitteeWitness {
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/lightClient/proofs.ts:6]: shouldn't this stay `allForks` to allow for bellatrix and future forks?

**Review by wemeetagain — DISMISSED** [review]

File: packages/beacon-state-transition/src/allForks/util/epochContext.ts [hunk 3]
```diff
@@ -570,6 +631,28 @@ export class EpochContext {
 
     this.effectiveBalanceIncrements[index] = Math.floor(effectiveBalance / EFFECTIVE_BALANCE_INCREMENT);
   }
+
+  /**
+   * Note: The range of slots a validator has to perform duties is off by one.
+   * The previous slot wording means that if your validator is in a sync committee for a period that runs from slot
+   * 100 to 200,then you would actually produce signatures in slot 99 - 199.
+   */
+  getIndexedSyncCommittee(slot: Slot): SyncCommitteeCache {
```
Comment (twoeths) [inline packages/beacon-state-transition/src/allForks/util/epochContext.ts:646]: Some consumers have to convert from epoch to slot, and this converts from slot to syncPeriod.
Should we separate this function to `getIndexedSyncCommitteeAtSlot()` and `getIndexedSyncCommitteeAtEpoch()`

File: packages/beacon-state-transition/src/allForks/util/epochContext.ts [hunk 4]
```diff
@@ -362,7 +379,36 @@ export class EpochContext {
    */
   exitQueueChurn: number;
 
-  constructor(data: IEpochContextData) {
+  /** TODO: Indexed SyncCommitteeCache */
```
Comment (twoeths) [inline packages/beacon-state-transition/src/allForks/util/epochContext.ts]: this SyncCommitteeCache is same to the old IndexedSyncCommittee, so drop this TODO and below?

File: packages/beacon-state-transition/src/allForks/util/epochContext.ts (hunk 3 above)
Comment (dapplion) [reply packages/beacon-state-transition/src/allForks/util/epochContext.ts:646]: Added getIndexedSyncCommitteeAtEpoch

File: packages/beacon-state-transition/src/allForks/util/epochContext.ts (hunk 4 above)
Comment (dapplion) [reply packages/beacon-state-transition/src/allForks/util/epochContext.ts]: Fixed

File: packages/lodestar/src/chain/lightClient/proofs.ts (hunk 2 above)
Comment (dapplion) [reply packages/lodestar/src/chain/lightClient/proofs.ts:6]: allForks doesn't have the fields for sync committees and this doesn't compile. The correct type would be allForks except phase0, but with current allforks types that's not possible. With SSZ refactor I can do that since allForks is purely a union

File: packages/beacon-state-transition/src/util/genesis.ts (hunk 1 above)
Comment (dapplion) [reply packages/beacon-state-transition/src/util/genesis.ts:100]: We need a CachedBeaconState to run processDeposit() which uses various caches. However at this point the state's syncCommittees are not known. This function can be called by:
  - 1. genesis spec tests: Don't care about the committee cache
  - 2. genesis builder: Only supports starting from genesis at phase0 fork
  - 3. interop state: Only supports starting from genesis at phase0 fork

So it's okay to skip syncing the sync committee cache here and expect it to be populated latter when the altair fork happens for cases 2, 3.

Comment (dapplion) [reply packages/beacon-state-transition/src/util/genesis.ts:100]: Note that in current master the sync committee caches are populated with wrong data, because the logic wasn't checking that the pubkeys in the state's syncCommittee actually existed

Comment (twoeths) [reply packages/beacon-state-transition/src/util/genesis.ts:100]: maybe we just need to drop a comment why we skip computing sync committees here, master works the same way (and it misses the comment that make me search for a while)

Comment (dapplion) [reply packages/beacon-state-transition/src/util/genesis.ts:100]: Already added the comment https://github.com/ChainSafe/lodestar/pull/3729/commits/7218fb011524b1cef43930f2adc6e8efe7305b7a

**Review by twoeths — APPROVED** [review]

**Comment by dapplion** [comment]
@wemeetagain I think all concerns are covered, and I'll merge to continue ssz refactor related PRs. If you have more concerns please comment and I can open fix PRs
